File: sdk/go/runs.go
package marmot

import (
	"context"

	apiclient "github.com/marmotdata/marmot/sdk/go/internal/gen/client"
	"github.com/marmotdata/marmot/sdk/go/internal/gen/client/runs"
	"github.com/marmotdata/marmot/sdk/go/internal/gen/models"
)

// Run is a single pipeline ingestion run.
type Run = models.PluginRun

// RunList is a paginated set of runs.
type RunList = runs.GetRunsOKBody

// RunEntities is the response from RunsService.Entities.
type RunEntities = models.V1RunsRunEntitiesResponse

// RunsListOptions filters RunsService.List. Pipelines and Statuses are comma-separated.
type RunsListOptions struct {
	Pipelines string
	Statuses  string
	Limit     int64
	Offset    int64
}

// RunEntitiesOptions filters RunsService.Entities.
type RunEntitiesOptions struct {
	EntityType string
	Status     string
	Limit      int64
	Offset     int64
}

// RunsService lists and inspects pipeline ingestion runs.
type RunsService struct {
	gen *apiclient.Marmot
}

// List returns paginated runs.
func (s *RunsService) List(ctx context.Context, opts RunsListOptions) (*RunList, error) {
	p := runs.NewGetRunsParams().WithContext(ctx)
	if opts.Pipelines != "" {
		p = p.WithPipelines(&opts.Pipelines)
	}
	if opts.Statuses != "" {
		p = p.WithStatuses(&opts.Statuses)
	}
	if opts.Limit > 0 {
		p = p.WithLimit(&opts.Limit)
	}
	if opts.Offset > 0 {
		p = p.WithOffset(&opts.Offset)
	}
	resp, err := s.gen.Runs.GetRuns(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// Get fetches a single run by ID.
func (s *RunsService) Get(ctx context.Context, id string) (*Run, error) {
	p := runs.NewGetRunsIDParams().WithContext(ctx).WithID(id)
	resp, err := s.gen.Runs.GetRunsID(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// Entities lists the entities processed in a run.
func (s *RunsService) Entities(ctx context.Context, runID string, opts RunEntitiesOptions) (*RunEntities, error) {
	p := runs.NewGetRunsIDEntitiesParams().WithContext(ctx).WithID(runID)
	if opts.EntityType != "" {
		p = p.WithEntityType(&opts.EntityType)
	}
	if opts.Status != "" {
		p = p.WithStatus(&opts.Status)
	}
	if opts.Limit > 0 {
		p = p.WithLimit(&opts.Limit)
	}
	if opts.Offset > 0 {
		p = p.WithOffset(&opts.Offset)
	}
	resp, err := s.gen.Runs.GetRunsIDEntities(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

File: sdk/python/src/marmot/resources/runs.py
"""List and inspect pipeline ingestion runs."""

from __future__ import annotations

from typing import cast

from marmot._adapter import unwrap
from marmot._gen.api.runs import get_runs, get_runs_id, get_runs_id_entities
from marmot._gen.client import AuthenticatedClient
from marmot._gen.models.get_runs_response_200 import GetRunsResponse200
from marmot._gen.models.plugin_run import PluginRun
from marmot._gen.models.run_entities_response import RunEntitiesResponse
from marmot._gen.types import UNSET, Unset


class RunsResource:
    def __init__(self, client: AuthenticatedClient) -> None:
        self._c = client

    def list(
        self,
        *,
        pipelines: str | None = None,
        statuses: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> GetRunsResponse200:
        """Return paginated runs. ``pipelines`` and ``statuses`` are comma-separated."""
        pipelines_arg: str | Unset = pipelines if pipelines is not None else UNSET
        statuses_arg: str | Unset = statuses if statuses is not None else UNSET
        limit_arg: int | Unset = limit if limit is not None else UNSET
        offset_arg: int | Unset = offset if offset is not None else UNSET
        return cast(
            GetRunsResponse200,
            unwrap(
                get_runs.sync_detailed(
                    client=self._c,
                    pipelines=pipelines_arg,
                    statuses=statuses_arg,
                    limit=limit_arg,
                    offset=offset_arg,
                )
            ),
        )

    def get(self, run_id: str) -> PluginRun:
        """Fetch a single run by its database ID."""
        return cast(PluginRun, unwrap(get_runs_id.sync_detailed(id=run_id, client=self._c)))

    def entities(
        self,
        run_id: str,
        *,
        entity_type: str | None = None,
        status: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> RunEntitiesResponse:
        """List the entities processed in a run."""
        et_arg: str | Unset = entity_type if entity_type is not None else UNSET
        status_arg: str | Unset = status if status is not None else UNSET
        limit_arg: int | Unset = limit if limit is not None else UNSET
        offset_arg: int | Unset = offset if offset is not None else UNSET
        return cast(
            RunEntitiesResponse,
            unwrap(
                get_runs_id_entities.sync_detailed(
                    id=run_id,
                    client=self._c,
                    entity_type=et_arg,
                    status=status_arg,
                    limit=limit_arg,
                    offset=offset_arg,
                )
            ),
        )
